### PR TITLE
Keep sessions in window for the full unfinalized chain

### PR DESCRIPTION
**Pull request by sandreim**
Fixes https://github.com/paritytech/polkadot/issues/6040

TODO:
- [x] fix and add tests
- [x] Versi burn-in

### CODE REVIEW

File: node/subsystem-util/src/rolling_session_window.rs [hunk 1]
```diff
@@ -146,19 +158,105 @@ impl RollingSessionWindow {
...
+		sender: &mut Sender,
 		block_hash: Hash,
-	) -> Result<SessionWindowUpdate, SessionsUnavailable> {
+	) -> Result<SessionWindowUpdate, SessionsUnavailable>
+	where
+		Sender: overseer::SubsystemSender<RuntimeApiMessage>
+			+ overseer::SubsystemSender<ChainApiMessage>,
+	{
 		let session_index = get_session_index_for_child(sender, block_hash).await?;
+		let last_finalized_block_session = self.last_finalized_block_session(sender).await?;
```
Comment (eskimor) [inline node/subsystem-util/src/rolling_session_window.rs:259]: Technically it is the earliest_non_finalized_block_session. (Which is exactly what we want)

**Review by eskimor — APPROVED** [review]

**Comment by sandreim** [comment]
dang, I need to fix tests in approval voting and dispute coordinator by handling the extra overseer traffic that is expected.

File: node/subsystem-util/src/rolling_session_window.rs [hunk 2]
```diff
@@ -146,19 +158,103 @@ impl RollingSessionWindow {
...
+					return Err(SessionsUnavailable {
+						kind: SessionsUnavailableReason::MissingLastFinalizedBlock,
+						info: None,
+					})
+				},
+			}
+		};
+
+		let (tx, rx) = oneshot::channel();
+		// We want to get the parent of the last finalized block.
```
Comment (vstakhov) [inline node/subsystem-util/src/rolling_session_window.rs:194]: I don't see where we get a parent here. It seems, that we extract the block hash at `last_finalized_height`, so either the comment (most likely) or the logic is wrong. For the parent of the last finalized block, we need to get a hash for `last_finalized_height - 1` I suppose, but from the following code, it seems that we actually need the `last_finalized_height` block hash.

File: node/subsystem-util/src/rolling_session_window.rs [hunk 3]
```diff
@@ -38,9 +40,18 @@ pub enum SessionsUnavailableReason {
...
+	ChainApi(#[from] ChainApiError),
 	/// Missing session info from runtime API for given `SessionIndex`.
 	#[error("Missing session index {0:?}")]
 	Missing(SessionIndex),
+	/// Missing last finalized block number.
+	#[error("Missing last finalized block number")]
+	MissingLastFinalizedBlock,
+	/// Missing last finalized block hash.
+	#[error("Missing last finalized block hash")]
+	MissingLastFinalizedBlockHash,
```
Comment (vstakhov) [inline node/subsystem-util/src/rolling_session_window.rs:54]: Maybe this error variant should have a parameter of a type `BlockNumber` to specify which block is missing?

**Review by vstakhov — APPROVED** [review]

File: node/subsystem-util/src/rolling_session_window.rs (hunk 2 above)
Comment (sandreim) [reply node/subsystem-util/src/rolling_session_window.rs:194]: Yeah, I have changed the code but forgot to do the same about then comment. The last finalized child session is what we are after here.

File: node/subsystem-util/src/rolling_session_window.rs (hunk 3 above)
Comment (sandreim) [reply node/subsystem-util/src/rolling_session_window.rs:54]: Good idea.

**Comment by sandreim** [comment]
Versi broke down during burn-in, cannot reason if this is related to the PR or just a deployment issue.

**Review by eskimor — APPROVED** [review]

**Pull request merged**